include/ensmallen_bits/nsgaiii/nsgaiii_impl.hpp

```diff
@@ -0,0 +1,314 @@
+
+inline NSGAIII::NSGAIII(const size_t populationSize,
+					    const size_t maxGenerations,
+					    const double crossoverProb,
+					    const double distrIndex) :
+		populationSize(populationSize),
+		maxGenerations(maxGenerations),
+		crossoverProb(crossoverProb),
+		distrIndex(distrIndex),
+		userDefinedSet(false)
+{ /* Nothing to do here */ }
+
+template<typename MultiObjectiveFunctionType>
+arma::cube NSGAIII::Optimize(MultiObjectiveFunctionType& function, arma::mat& iterate)
+{
+	// Initialize initial population.
+	arma::cube population = arma::randn(iterate.n_rows, iterate.n_cols, populationSize);
+	population.each_slice() += iterate;
+
+	// Initialize helper variables.
+	std::vector<size_t> S;
+	arma::cube R, Q;
+	std::vector<std::vector<size_t>> fronts;
+
+	for (size_t gen = 0; gen < maxGenerations; gen++)
+	{
+	  // Mate and create offsprings.
+	  Q = Mate(population);
+
+	  // Create a population made of the old population and the new offsprings.
+	  R = arma::join_slices(population, Q);
+
+	  // Evaluate R.
+	  arma::mat fitnessValues(function.NumObjectives(), R.n_slices);
+	  for (size_t i = 0; i < R.n_slices; i++)
+		fitnessValues.col(i) = function.Evaluate(R.slice(i));
+
+	  // Find next population.
+	  fronts.clear();
+	  NonDominatedSorting(fitnessValues, fronts);
+	  size_t l = 0, numMembers = 0;
+	  arma::cube nextPop = population;
+	  arma::mat newFitness(function.NumObjectives(), populationSize);
+	  while (numMembers < populationSize)
+	  {
+	    if (fronts[l].size() + numMembers >= populationSize)
+	      break;
+	    for (size_t i = 0; i < fronts[l].size(); i++)
+	    {
+		  newFitness.col(numMembers) = fitnessValues.col(fronts[l][i]);
+		  nextPop.slice(numMembers++) = population.slice(fronts[l][i]);
+	    }
+	    l++;
+	  }
+	  size_t K = population - numMembers;
+	  if (K == 0)
+	  	continue;
+
+	  // Normalization step.
+	  arma::mat asf(function.NumObjectives(), function.NumObjectives(),
+		  arma::fill::eye);
+
+	  for (size_t i = 0; i < asf.n_rows; i++)
+	  {
+	    for (size_t j = 0; j < asf.n_cols; j++)
+	    {
+		  if (asf(i, j) == 0)
+		    asf(i, j) = 1e6;
+	    }
+	  }
+	
+	  arma::mat f(function.NumObjectives(), fronts[0].size());
+	  for (size_t i = 0; i < f.n_cols; i++)
+	    f.col(i) = fitnessValues.col(fronts[0][i]);
+
+	  arma::vec zmin(function.NumObjectives());
+	  for (size_t i = 0; i < function.NumObjectives(); i++)
+	    zmin[i] = f.row(i).min();
+
+	  for (size_t i = 0; i < fronts[0].size(); i++)
+	    f.col(i) -= zmin;
+
+	  for (auto it = fitnessValues.begin(); it != fitnessValues.end(); it++)
+	    if (*it <= 1e-3)
+		  *it = 0;
+
+	  arma::mat max(asf.n_rows, fronts[0].size());
+	  for (size_t i = 0; i < asf.n_rows; i++)
+	  {
+	    arma::mat f_asf = f.t() * asf.row(i);
+	    for (size_t j = 0; j < f_asf.n_rows; j++)
+		  max(i, j) = f_asf.row(j).max();
+	  }
+
+	  arma::vec mins(max.n_rows);
+	  for (size_t i = 0; i < mins.n_elem; i++)
+	    mins[i] = max.row(i).index_max();
```

Comment:
We can bump the armadillo version as done here: https://github.com/mlpack/ensmallen/pull/119/files#diff-354f30a63fb0907d4ad57269548329e3 to fix the build issue.